plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,472 @@
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_magic_unlock(FuCcgxPureHidDevice *self, GError **error)
+{
+	guint8 buf[8] = {FU_CCGX_PURE_HID_REPORT_ID_CUSTOM,
+			 FU_CCGX_PD_RESP_BRIDGE_MODE_CMD_SIG,
+			 0x43,
+			 0x59,
+			 0x00,
+			 0x00,
+			 0x00,
+			 0x0B};
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error)) {
+		g_prefix_error(error, "magic enable command error: ");
+		return FALSE;
+	}
+
+	/* ignore error: this always fails but has the correct behavior */
+	if (!fu_ccgx_pure_hid_command(self,
+				      FU_CCGX_PURE_HID_COMMAND_MODE,
+				      FU_CCGX_PD_RESP_BRIDGE_MODE_CMD_SIG,
+				      &error_local)) {
+		g_debug("expected HID report bridge mode failure: %s", error_local->message);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_ensure_fw_info(FuCcgxPureHidDevice *self, GError **error)
+{
+	FuDevice *device = FU_DEVICE(self);
+	guint8 buf[0x40] = {FU_CCGX_PURE_HID_REPORT_ID_INFO, 0};
+	guint version = 0;
+	g_autofree gchar *bl_ver = NULL;
+	g_autoptr(GByteArray) st_info = NULL;
+
+	if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error))
+		return FALSE;
+
+	st_info = fu_struct_ccgx_pure_hid_fw_info_parse(buf, sizeof(buf), 0x0, error);
+	if (st_info == NULL)
+		return FALSE;
+	self->silicon_id = fu_struct_ccgx_pure_hid_fw_info_get_silicon_id(st_info);
+	self->operating_mode = fu_struct_ccgx_pure_hid_fw_info_get_operating_mode(st_info);
+
+	fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	/* set current version */
+	switch (self->operating_mode) {
+	case FU_CCGX_PURE_HID_FW_MODE_FW1:
+		version = fu_struct_ccgx_pure_hid_fw_info_get_image1_version(st_info);
+		break;
+	case FU_CCGX_PURE_HID_FW_MODE_FW2:
+		version = fu_struct_ccgx_pure_hid_fw_info_get_image2_version(st_info);
+		break;
+	case FU_CCGX_PURE_HID_FW_MODE_BOOT:
+		/* force an upgrade to any version */
+		version = 0x0;
+		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+		break;
+	default:
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "unsupported mode");
+		return FALSE;
+	}
+	fu_device_set_version_raw(device, version);
+
+	/* set bootloader version */
+	fu_device_set_version_bootloader_raw(
+	    device,
+	    fu_struct_ccgx_pure_hid_fw_info_get_bl_version(st_info));
+	bl_ver = fu_version_from_uint32(fu_struct_ccgx_pure_hid_fw_info_get_bl_version(st_info),
+					fu_device_get_version_format(self));
+	fu_device_set_version_bootloader(FU_DEVICE(self), bl_ver);
+
+	/* FIXME: d4s: do we need that for querying??? */
+	if (!fu_ccgx_pure_hid_enter_flashing_mode(self, error))
+		return FALSE;
```

Comment:
Can you try to drop this?  It's better we don't swap modes every time the daemon starts up IMO.

Reply:
According to Wireshark, it is used for querying with Windows utility from Framework.
However the detection works without it.
I moved it to `prepare_firmware()` since I don't want to drop this command from the write sequence and I have no any documentation to check if it is really required.

Reply:
Yeah it makes sense in the write sequence, but don't think we should have it otherwise.

BTW - since you captured with wireshark in Windows, can you do device installation emulation data too?  We can add it to the installed tests then to keep this thing working.